src/common/arrow/schema_metadata.cpp

```diff
@@ -48,14 +50,40 @@ string ArrowSchemaMetadata::GetExtensionName() const {
 	return GetOption(ARROW_EXTENSION_NAME);
 }
 
-ArrowSchemaMetadata ArrowSchemaMetadata::MetadataFromName(const string &extension_name) {
+ArrowSchemaMetadata ArrowSchemaMetadata::ArrowCanonicalType(const string &extension_name) {
+	ArrowSchemaMetadata metadata;
+	metadata.AddOption(ARROW_EXTENSION_NAME, extension_name);
+	metadata.AddOption(ARROW_METADATA_KEY, "");
+	return metadata;
+}
+
+ArrowSchemaMetadata ArrowSchemaMetadata::DuckDBInternalType(const string &type_name) {
 	ArrowSchemaMetadata metadata;
-	metadata.AddOption(ArrowSchemaMetadata::ARROW_EXTENSION_NAME, extension_name);
-	metadata.AddOption(ArrowSchemaMetadata::ARROW_METADATA_KEY, "");
+	metadata.AddOption(ARROW_EXTENSION_NAME, ARROW_EXTENSION_NON_CANONICAL);
+	// We have to set the metadata key with type_name and vendor_name.
+	metadata.extension_metadata_map["vendor_name"] = "DuckDB";
+	metadata.extension_metadata_map["type_name"] = type_name;
+	metadata.AddOption(ARROW_METADATA_KEY, StringUtil::ToJSONMap(metadata.extension_metadata_map));
 	return metadata;
 }
 
-bool ArrowSchemaMetadata::HasExtension() {
+bool ArrowSchemaMetadata::IsNonCanonicalType(const string &type, const string &vendor) const {
+	if (schema_metadata_map.find(ARROW_EXTENSION_NAME) == schema_metadata_map.end()) {
+		return false;
+	}
+	if (schema_metadata_map.find(ARROW_EXTENSION_NAME)->second != "arrow.opaque") {
```

Comment:
nit: Can/should this use `ARROW_EXTENSION_NON_CANONICAL` instead of `"arrow.opaque"`?